docs/non-functional-requirements/performance.md

```diff
@@ -0,0 +1,34 @@
+# Performance
+
+Performance refers to the responsiveness, efficiency, and speed with which a system completes tasks and processes user requests. It encompasses several key metrics such as response time, throughput, latency, and resource utilization.
+
+## Characteristics
+
+- Response Time: The time taken by the system to respond to user interactions or requests. Lower response times indicate better performance and user responsiveness.
+- Throughput: The rate at which the system can process and handle a certain volume of transactions or requests within a given time frame. Higher throughput signifies greater processing capacity and efficiency.
+- Latency: The delay or time lag experienced between initiating a request and receiving a response. Low latency is crucial for real-time applications to ensure timely interactions.
+- Scalability: The system's ability to handle increasing workload or user demand by scaling resources (horizontal or vertical scaling) without impacting performance negatively.
+- Concurrency: The system's capability to handle multiple concurrent users or tasks efficiently without significant degradation in performance. This involves managing resources such as CPU, memory, and network bandwidth effectively.
+- Resource Utilization: Efficient utilization of hardware resources (e.g., CPU, memory, disk) to maximize performance without unnecessary overhead or bottlenecks.
+- Stability: Consistency and reliability of performance over time and under varying conditions, ensuring predictable behavior and minimal downtime.
+- Fault Tolerance: The system's ability to continue operating or recover gracefully from failures or disruptions without significant impact on performance or user experience.
+- Load Handling: How well the system manages and distributes workload during peak usage periods to maintain optimal performance levels.
+
+## Implementations
+
+Implementing performance involves a combination of architectural decisions, coding practices, infrastructure setup, and optimization techniques. For example:
+
+- Efficient Algorithms and Data Structures: Choosing algorithms and data structures that are optimized for the specific tasks and operations performed by the system can significantly improve performance. This includes selecting algorithms with lower time complexity (e.g., O(1), O(log n)) for critical operations.
+- Code Optimization: Writing efficient and optimized code reduces execution time and resource consumption. Techniques such as minimizing loops, reducing unnecessary computations, and using appropriate data types can improve performance.
+- Concurrency and Parallelism: Leveraging concurrency models (e.g., threads, async-await) and parallel programming techniques allows the system to execute multiple tasks simultaneously, utilizing multi-core processors effectively and improving throughput.
```

Comment:
I would split this up to 
* Concurrency 
* Parallelism 
As people tend to confuse between the two and think they are the same but not sure it is really needed here.
Your call.